File: leetcode/1025_divisor_game.cpp
// Copyright (c) 2019 Jacopo Notarstefano

#include <cassert>
#include <vector>

using namespace std;

class Solution {
 public:
    bool divisorGame(int N) {
        vector<bool> result(N + 1);

        for (int i = 2; i <= N; i++) {
            for (int j = 1; j < i; j++) {
                if (i % j == 0 && !result[i - j]) {
                    result[i] = true;
                }
            }
        }

        return result[N];
    }
};

int main() {
    auto solution = Solution();

    assert(solution.divisorGame(2));
    assert(!solution.divisorGame(3));
}

File: leetcode/1025_divisor_game.py
# -*- coding: utf-8 -*-


class Solution:
    def divisorGame(self, N):
        result = [False] * (N + 1)

        for i in range(2, N + 1):
            for j in range(1, i):
                if i % j == 0 and not result[i - j]:
                    result[i] = True

        return result[N]


if __name__ == '__main__':
    solution = Solution()

    assert solution.divisorGame(2)
    assert not solution.divisorGame(3)
